Change representation of signed integer values to use sign extension

jafingerhut opened the pull request.

jafingerhut commented on docs/v1/P4Runtime-Spec.mdk:
```diff
@@ -1649,53 +1649,85 @@ fits the encoded integer value. This representation achieves three goals:
-  encoded_string_size = (high_bit + 8) / 8;
-else
-  encoded_string_size = 1;
+encoded_string_size = floor((A + 7) / 8)
```
From teverman: Would it be clearer to rename encoded_string_size to minimum_string_size? Just a thought.

jafingerhut replied:
Done.

jafingerhut commented on docs/v1/P4Runtime-Spec.mdk:
```diff
@@ -1649,53 +1649,85 @@ fits the encoded integer value. This representation achieves three goals:
+
+For received values expected to fit within a `bit<W>` type, the value
+is in range if, after removing all most significant 0 bits, the value
+remaining is `W` bits or less.
```
From teverman: I think it's more accurate to specify the value width here, i.e "the remaining value's width is W bits or less".

jafingerhut replied:
Done.

jafingerhut commented on docs/v1/P4Runtime-Spec.mdk:
```diff
@@ -1649,53 +1649,85 @@ fits the encoded integer value. This representation achieves three goals:
+
+For received values expected to fit within an `int<W>` type, the value
+is in range if, after "undoing sign extension", the value remaining is
+`W` bits or less. To undo sign extension, start by eliminating the
```
From teverman: Add width per previous comment.

jafingerhut replied:
Done.

antoninbas commented on docs/v1/P4Runtime-Spec.mdk:
```diff
@@ -1649,53 +1649,85 @@ fits the encoded integer value. This representation achieves three goals:
+client can continue to transmit P4Runtime messages between them when
+one has a P4Info file for version A of a P4 program, at the same time
+that the other has a P4Info file for version B of a P4 program, and
+those P4 programs differ in the bit widths of some values of type
```
I think the rest of the spec uses the "bitwidth" syntax

jafingerhut replied:
Done in the new branch I created in this repo: https://github.com/p4lang/p4runtime/tree/issue111-change-signed-int-representation

antoninbas commented on docs/v1/P4Runtime-Spec.mdk:
```diff
@@ -1649,53 +1649,85 @@ fits the encoded integer value. This representation achieves three goals:
+versa.  If you attempt to do so, this mechanism can quietly change
+negative signed values to positive unsigned values, or vice versa.  It
+also limits the magnitude of the values transmitted to those that fit
+within the smaller of the bit widths supported by either end of the
```
same syntax comment as above

jafingerhut replied:
Done in the new branch I created in this repo: https://github.com/p4lang/p4runtime/tree/issue111-change-signed-int-representation

antoninbas commented on docs/v1/P4Runtime-Spec.mdk:
```diff
@@ -1649,53 +1649,85 @@ fits the encoded integer value. This representation achieves three goals:
+
+For received values expected to fit within an `int<W>` type, the value
+is in range if, after "undoing sign extension", the remaining value's
+width is `W` bits or less. To undo sign extension, start by
```
I may be wrong, but isn't it `W-1` bits here?
For example, if I expect `int<2>` (only 1 and -1 can be represented) and the client sends the 1-byte value `00000011b`, then after "undoing sign extension" I get a value (3) that fits in 2 bits but cannot be represented with `int<2>`.

jafingerhut replied:
The type `int<W>` can always represent `2^W` different values, e.g. `int<2>` can represent any value in the range `[-2, 1]`.

A P4Runtime client/server, if it started with a value that it thought was type `int<2>` represented as `11b`, would _never_ send it as the byte `00000011b` using the proposed mechanism.  It would sign extend the 2-bit value with 1 bits, and send it as `11111111b`.  Both `11b` and `11111111b` represent the numeric value -1 in 2's complement representation.

The only way with the proposed changes that one of the client/server would send `00000011b` as the encoding for a value of type `int<W>` is if the numeric value was +3, and the type would thus have to be `int<3>` or larger at the sender.   If the receiver had type `int<2>` or `int<1>`, i would "un sign extend" the value down to `011b`, but would there have to stop because the 2 most significant bits differ, and reducing it more would change the sign of the value.  It would then detect that the value requires a width of W=3 bits, and since the receiver's type was `int<2>` or `int<1>` in this example scenario, the receiver would declare the value out of range.

jafingerhut replied:
I can add example scenarios like this, either directly in this section explaining this mechanism, or in an appendix, if you think having such examples would clarify things for readers.  I know that many are not as intimately familiar with the gory details of 2's complement notation as I forced myself to become :)

antoninbas replied:
Yes I did forget that P4 used 2's complement for a second there :)
I was also thinking of a sender with `int<3>` and a receiver with `int<2>`, but didn't "unsign extend" properly. My only suggestion would be to maybe use binary string when talking about the bit pattern sent by the client and value for the numerical value. Apart from that I think it's clear.

jafingerhut replied:
I have made a few changes from "value" to "bit string" in the version pushed to this branch: https://github.com/p4lang/p4runtime/tree/issue111-change-signed-int-representation

jafingerhut commented:
I created and committed a new branch in this repository: Done in the new branch I created in this repo: https://github.com/p4lang/p4runtime/tree/issue111-change-signed-int-representation

Is there some way I should be able to create  PR from this branch?  I don't see a way to do that in Github's web interface yet.

antoninbas commented:
I have reviewed the generated outputs and they look good
One minor nit: the rest of the spec uses "bitstring" consistently, not "bit string"
I think that the only reason why you don't see an option to create a PR is because the branch is one commit behind master. If you rebase master into your branch, you should see the option when you visit: https://github.com/p4lang/p4runtime/tree/issue111-change-signed-int-representation

jafingerhut commented:
I have replaced a couple of occurrences of "bit string" with "bitstring" for consistency with the rest of the document, rebased the changes on top of the latest master, and created this PR: https://github.com/p4lang/p4runtime/pull/152

antoninbas commented:
I'm closing this as it was superseded by #152
